AppController/djinn.rb

```diff
@@ -1922,37 +1936,48 @@ def job_start(secret)
       # the state in case we are looking for a zookeeper server.
       pick_zookeeper(@zookeeper_data)
 
-      @state = "Done starting up AppScale, now in heartbeat mode"
-
       # We save the current @options and roles to check if
       # restore_appcontroller_state modifies them.
       old_options = @options.clone
       old_jobs = my_node.jobs
 
       # The following is the core of the duty cycle: start new apps,
       # restart apps, terminate non-responsive AppServers, and autoscale.
-      # Every other node syncs its state with the login node state.
+
+      # Every other node syncs its state with the login node state. The
+      # load_balancers needs to check the applications that got loaded
+      # this time, to setup the routining.
+      my_versions_loaded = @versions_loaded if my_node.is_load_balancer?
       if my_node.is_shadow?
         update_node_info_cache
         backup_appcontroller_state
+      elsif !restore_appcontroller_state
+        @state = "Couldn't reach the deployment state: now in isolated mode"
+        Djinn.log_warn("Cannot talk to zookeeper: in isolated mode.")
+        next
+      end
 
+      # We act here if options or roles for this node changed.
+      check_role_change(old_options, old_jobs)
+
+      # Load balancers (and shadow) needs to setup new applications.
+      if my_node.is_load_balancer?
         APPS_LOCK.synchronize {
           # Starts apps that are not running yet but they should.
-          versions_to_load = ZKInterface.get_versions - @versions_loaded
+          if my_node.is_shadow?
+            versions_to_load = ZKInterface.get_versions - @versions_loaded
+          else
+            versions_to_load = @versions_loaded - my_versions_loaded
```

Comment:
aren't `my_versions_loaded` and `@versions_loaded` the same

Reply:
the restore_appcontroller_state few lines above, reloads @versions_loaded from what the master sees, so, they are not the same when applications get deployed.

Reply:
gotcha